### PR TITLE
Tg 27 implement generator loss

### PR DESCRIPTION
**Pull request by rohit901**
@Dmmc123 review please.

### CODE REVIEW

File: src/models/losses.py [hunk 1]
```diff
@@ -4,6 +4,327 @@
...
+        []
+    )  # Mask mis-match samples, that come from the same class as the real sample ###
+
+    match_scores = []
+    cap_len_list = cap_lens.data.tolist()
+    gamma1 = const_dict["gamma1"]
+    gamma2 = const_dict["gamma2"]
+    gamma3 = const_dict["gamma3"]
+
+    for i in range(batch_size):
```
Comment (Dmmc123) [inline src/models/losses.py:147]: I understand that it's easier to make it a loop and process each sample individually however I think it's beneficial for performance to try to use the vectorized operations on batch tensors

slow:
```
result = []
for i in range(t.size(0)):
    result.append(op(t[i]))
```

fast:
```
result = t.op()
```

Can you give it a thought?

File: src/models/losses.py [hunk 2]
```diff
@@ -4,6 +4,327 @@
...
+        ] = 0  # This ensures that "correct class" index is not included in the mask.
+        masks.append(mask.reshape(1, -1))  # shape: (1, batch)
+
+        numb_words = cap_len_list[i]
+        query_words = (
+            words_emb[i, :, :numb_words].unsqueeze(0).contiguous()
+        )  # shape: (1, D, L), this picks the caption at ith batch index.
+        query_words = query_words.repeat(
+            batch_size, 1, 1
+        )  # shape: (batch, D, L), this expands the same caption for all batch indices.
```
Comment (Dmmc123) [inline src/models/losses.py:160]: Why you expand same caption for all batch indices? I thought 1 caption for 1 sample from batch, no?

**Review by Dmmc123 — CHANGES_REQUESTED** [review]

File: src/models/losses.py [hunk 3]
```diff
@@ -4,6 +4,327 @@
...
+        this is a tensor of ones. [shape: (batch_size, 1)].
+
+        words_emb: The embeddings for all the words in the captions.
+        shape: (batch_size, embedding_size, max_caption_length)
+
+        sent_emb: The embeddings for the sentences.
+        shape: (batch_size, embedding_size)
+
+        match_labels: Tensor of shape: (batch_size, 1).
+        This is of the form torch.tensor([0, 1, 2, ..., batch-1])
```
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:45]: true, but all the labels [real_labels, fake_labels, match_labels] are handled by a separate function called prepare_labels, and it does not make sense to generate labels inside this discriminator loss, instead we generate at one central place and pass it in the training loop to different places

File: src/models/losses.py [hunk 4]
```diff
@@ -4,6 +4,327 @@
...
+        cap_lens: The length of the 'actual' captions in the batch [without padding]
+        shape: (batch_size, 1)
+
+        class_ids: The class ids for the instance. shape: (batch_size, 1)
+
+        vgg_encoder: The VGG encoder model.
+
+        real_imgs: The Real Images from the dataset. shape: (batch_size, 3, 256, 256)
+
+        device: The device to run the model on.
```
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:56]: device is being used to transfer the mask tensor created inside DAMSM loss into GPU. Since this tensor is created inside this loss for purpose of this loss, might have to pass in device if we want that tensor to be in GPU, if not can get rid of the device argument but then the tensor operation will be done in CPU which might be slow. So what do you think.

File: src/models/losses.py [hunk 5]
```diff
@@ -4,6 +4,327 @@
...
+    real_imgs: torch.Tensor,
+    device: Any,
+    const_dict: dict[Any, Any],
+) -> Any:
+    """Calculate the loss for the generator.
+
+    Args:
+        logits: Dictionary with fake/real and word-level/uncond/cond logits
+
+        inception_encoder: The inception encoder model.
```
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:31]: Actually yes I thought about it too, but then currently in the main training loop we pass real images to inception encoder and not fake images. Fake images are passed only for loss calculation purposes and hence I have used them here. I believe this might make the training loop a bit clear and clean and would be better in understanding the flow. I feel the current code is abstracting out the loss calculation by doing it inside the loss function. Let me know if you think this is bad

File: src/models/losses.py [hunk 6]
```diff
@@ -4,6 +4,327 @@
...
+
+        match_labels: Tensor of shape: (batch_size, 1).
+        This is of the form torch.tensor([0, 1, 2, ..., batch-1])
+
+        cap_lens: The length of the 'actual' captions in the batch [without padding]
+        shape: (batch_size, 1)
+
+        class_ids: The class ids for the instance. shape: (batch_size, 1)
+
+        vgg_encoder: The VGG encoder model.
```
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:52]: vgg encoder is used for perceptual loss calculation where we also pass fake images to the encoder. passing it here as features of fake image from VGG encoder is only used for perceptual loss, felt it would abstract the loss more compared to using in training loop. let me know your thoughts.

File: src/models/losses.py [hunk 7]
```diff
@@ -4,6 +4,327 @@
...
+    real_labels: torch.Tensor,
+    words_emb: torch.Tensor,
+    sent_emb: torch.Tensor,
+    match_labels: torch.Tensor,
+    cap_lens: torch.Tensor,
+    class_ids: torch.Tensor,
+    vgg_encoder: Any,
+    real_imgs: torch.Tensor,
+    device: Any,
+    const_dict: dict[Any, Any],
```
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:24]: changed it

File: src/models/losses.py [hunk 8]
```diff
@@ -4,6 +4,327 @@
...
+
+    match_scores = []
+    cap_len_list = cap_lens.data.tolist()
+    gamma1 = const_dict["gamma1"]
+    gamma2 = const_dict["gamma2"]
+    gamma3 = const_dict["gamma3"]
+
+    for i in range(batch_size):
+        mask = (class_ids == class_ids[i]).type(torch.int)
+        mask[
```
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:149]: made the change

File: src/models/losses.py [hunk 9]
```diff
@@ -4,6 +4,327 @@
...
+    fake_imgs: torch.Tensor,
+    real_labels: torch.Tensor,
+    words_emb: torch.Tensor,
+    sent_emb: torch.Tensor,
+    match_labels: torch.Tensor,
+    cap_lens: torch.Tensor,
+    class_ids: torch.Tensor,
+    vgg_encoder: Any,
+    real_imgs: torch.Tensor,
+    device: Any,
```
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:23]: yeah i made it thinking that we would be passing torch.cuda.device type and not a string here. so i guess this can remain as it is?

File: src/models/losses.py [hunk 10]
```diff
@@ -4,6 +4,327 @@
 import torch
 from torch import nn
 
+# pylint: disable=too-many-arguments
+# pylint: disable=too-many-locals
+
+
+def generator_loss(
+    logits: dict[str, dict[str, torch.Tensor]],
+    inception_encoder: Any,
```
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:13]: will make this change based on the result of previous discussion once you reply/resolve them.

File: src/models/losses.py [hunk 11]
```diff
@@ -4,6 +4,327 @@
...
+    fake_imgs: torch.Tensor,
+    real_labels: torch.Tensor,
+    words_emb: torch.Tensor,
+    sent_emb: torch.Tensor,
+    match_labels: torch.Tensor,
+    cap_lens: torch.Tensor,
+    class_ids: torch.Tensor,
+    vgg_encoder: Any,
```
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:21]: will make this change based on the result of previous discussion once you reply/resolve them.

File: src/models/losses.py [hunk 12]
```diff
@@ -4,6 +4,327 @@
...
+    alpha_j = alpha_j.view(batch, L, N)  # shape: (batch, L, N)
+    alpha_j_t = torch.transpose(alpha_j, 1, 2)  # shape: (batch, N, L)
+
+    c_i = (
+        local_incept_feat @ alpha_j_t
+    )  # shape: (batch, D, L) [summing over N dimension in paper, so we multiply like this]
+    return c_i
+
+
+def mean_squared_error(input_tensor: torch.Tensor, target: torch.Tensor) -> Any:
```
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:296]: done these set of changes

File: src/models/losses.py [hunk 13]
```diff
@@ -4,6 +4,327 @@
...
+    vgg_fake = vgg_encoder(fake_imgs)  # shape: (batch, 128, 128, 128)
+
+    loss_per = mean_squared_error(vgg_real, vgg_fake)  # perceptual loss
+
+    total_error_g += loss_per / 2.0
+    return total_error_g
+
+
+def damsm_loss(
+    local_incept_feat: torch.Tensor,
```
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:101]: yes the features corresponding to fake image obtained from inception is used only by damsm loss, so the logic for calculation of damsm is in that function while we pass the required inputs using the inception encoder which is passed to G

File: src/models/losses.py [hunk 14]
```diff
@@ -4,6 +4,327 @@
...
+    global_incept_feat: torch.Tensor,
+    words_emb: torch.Tensor,
+    sent_emb: torch.Tensor,
+    match_labels: torch.Tensor,
```
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:105]: same reply as earlier, it is initialized at central place prepare_labels and passed at one single place.

File: src/models/losses.py [hunk 15]
```diff
@@ -4,6 +4,327 @@
...
+    cap_lens: torch.Tensor,
+    class_ids: torch.Tensor,
+    device: Any,
```
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:108]: we are using pytorch device type, so using "Any", is there any other specific type for it that you want me to use?
torch.device("cuda")?

File: src/models/losses.py [hunk 16]
```diff
@@ -4,6 +4,327 @@
...
+    const_dict: dict[Any, Any],
```
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:109]: made this change

File: src/models/losses.py [hunk 17]
```diff
@@ -4,6 +4,327 @@
...
+
+        match_labels: Tensor of shape: (batch_size, 1).
+        This is of the form torch.tensor([0, 1, 2, ..., batch-1])
+
+        cap_lens: The length of the 'actual' captions in the batch [without padding]
+        shape: (batch_size, 1)
+
+        class_ids: The class ids for the instance. shape: (batch, 1)
+
+        device: The device to run the model on.
```
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:132]: used to transfer masks to GPU..

File: src/models/losses.py [hunk 18]
```diff
@@ -4,6 +4,327 @@
...
+    )  # Mask mis-match samples, that come from the same class as the real sample ###
+
+    match_scores = []
+    cap_len_list = cap_lens.data.tolist()
+    gamma1 = const_dict["gamma1"]
+    gamma2 = const_dict["gamma2"]
+    gamma3 = const_dict["gamma3"]
+
+    for i in range(batch_size):
+        mask = (class_ids == class_ids[i]).type(torch.int)
```
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:148]: done

File: src/models/losses.py [hunk 19]
```diff
@@ -4,6 +4,327 @@
...
+        shape: (batch_size, 1)
+
+        class_ids: The class ids for the instance. shape: (batch, 1)
+
+        device: The device to run the model on.
+
+        const_dict: The dictionary containing the constants.
+    """
+    batch_size = match_labels.size(0)
+    masks = (
```
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:137]: its a list, i guess due to comment the formatter screwed it up lol, setting the comment to previous line.

File: src/models/losses.py [hunk 20]
```diff
@@ -4,6 +4,327 @@
...
+    global_match_t = global_match_score.transpose(0, 1)  # shape: (batch, batch)
+
+    l1_s = cross_entropy(
+        global_match_score, match_labels
+    )  # This corresponds to L1_s from attnGAN.
+    l2_s = cross_entropy(
+        global_match_t, match_labels
+    )  # This corresponds to L2_s from attnGAN.
+
+    return l1_w, l2_w, l1_s, l2_s
```
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:236]: fixed it

File: src/models/losses.py [hunk 21]
```diff
@@ -4,6 +4,327 @@
...
+    cond_logits = logits["fake"]["cond"]
+    cond_err_g = binary_cross_entropy(cond_logits, real_labels)
+
+    uncond_logits = logits["fake"]["uncond"]
+    uncond_err_g = binary_cross_entropy(uncond_logits, real_labels)
+
+    loss_g = (
+        cond_err_g + uncond_err_g
+    )  # add up the conditional and unconditional losses
+    total_error_g += loss_g
```
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:72]: fixed this

File: src/models/losses.py [hunk 22]
```diff
@@ -4,6 +4,327 @@
...
+    loss_damsm = lambda3 * (l1_w + l2_w + l1_s + l2_s)
+
+    total_error_g += loss_damsm
+
+    vgg_real = vgg_encoder(real_imgs)  # shape: (batch, 128, 128, 128)
+    vgg_fake = vgg_encoder(fake_imgs)  # shape: (batch, 128, 128, 128)
+
+    loss_per = mean_squared_error(vgg_real, vgg_fake)  # perceptual loss
+
+    total_error_g += loss_per / 2.0
```
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:96]: ok fixed this

File: src/models/losses.py [hunk 23]
```diff
@@ -4,6 +4,327 @@
...
+    total_error_g = 0
+    lambda3 = const_dict["lambda3"]
+
+    cond_logits = logits["fake"]["cond"]
+    cond_err_g = binary_cross_entropy(cond_logits, real_labels)
+
+    uncond_logits = logits["fake"]["uncond"]
+    uncond_err_g = binary_cross_entropy(uncond_logits, real_labels)
+
+    loss_g = (
```
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:69]: blame the linter xD

[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:69]: anyways it should be fixed hopefully

File: src/models/losses.py [hunk 24]
```diff
@@ -4,6 +4,327 @@
...
+
+        const_dict: The dictionary containing the constants.
+    """
+    batch_size = match_labels.size(0)
+    masks = (
+        []
+    )  # Mask mis-match samples, that come from the same class as the real sample ###
+
+    match_scores = []
+    cap_len_list = cap_lens.data.tolist()
```
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:142]: possible with tensor but need to use .item() right. i think both are equivalent and we can go with anyone, hope nothing wrong with current implementation.

File: src/models/losses.py [hunk 25]
```diff
@@ -4,6 +4,327 @@
...
+    for i in range(batch_size):
+        mask = (class_ids == class_ids[i]).type(torch.int)
+        mask[
+            i
+        ] = 0  # This ensures that "correct class" index is not included in the mask.
+        masks.append(mask.reshape(1, -1))  # shape: (1, batch)
+
+        numb_words = cap_len_list[i]
+        query_words = (
+            words_emb[i, :, :numb_words].unsqueeze(0).contiguous()
```
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:156]: i see.. yeah i guess felt that after transformation [transpose, unsqueeze, etc] the mapping from index to memory would not be the same so thought this might cause some issues... anyway will remove it for now..

File: src/models/losses.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Dmmc123) [reply src/models/losses.py:56]: usually losses are supposed to be purely functional and there's almost no difference between running on GPU or CPU so loss is usually made unaware of device

if operations that involve `mask` operate on big tensors and can be a potential bottleneck then ok, it makes sense to make loss aware of device and do calculations on GPU sometimes

but if you think that those operations don't contribute much to the overall time, then I would suggest to remove the device dependancy and calculate everything on CPU

File: src/models/losses.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Dmmc123) [reply src/models/losses.py:31]: again, the idea of loss is essentially like this:
```
def my_awesome_loss(net_out, true_out):
    # some functional op on tensors
    return some_num
```
and in general while training loop you do
```
for epoch in range(epochs):
    net_out = net(x)
    loss = my_awesome_loss(net_out, true_out)
```
notice how you don't pass the network into loss because the moment you make loss dependent on the particular net, it introduces high coupling which is what we don't want

File: src/models/losses.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Dmmc123) [reply src/models/losses.py:52]: > felt it would abstract the loss

Which is exactly what we want. Think of default pytorch losses, they don'y take networks as inputs, they take outputs and what they should be. Losses shouldn't be aware of the flow of system, they're pure mathematical functions. There's a reason why we make loss functions a separate functions and don't calculate that entirely in the training loop and that reason is that it helps **reduce coupling** between net inference and net evaluation

File: src/models/losses.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Dmmc123) [reply src/models/losses.py:23]: Currently your device is said to have a type of `Any` and if you want to use `torch.cuda.device`, you better use this type

File: src/models/losses.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (Dmmc123) [reply src/models/losses.py:108]: `torch.cuda.device` or `str` will work just fine

File: src/models/losses.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (Dmmc123) [reply src/models/losses.py:142]: There's no need for `.item()` when indexing so this line is still useless
image

File: src/models/losses.py (hunk 1 above)
Comment (rohit901) [reply src/models/losses.py:147]: agree with you, but would prefer to do it later. Feel like not a top priority at the moment [and also bit busy with other projects and work]. Let's get started with results/experiments for now.

File: src/models/losses.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (Dmmc123) [reply src/models/losses.py:156]: > mapping from index to memory would not be the same

If that's the case then I think you need to specify the desired shapes and if it doesn't make a difference then I would omit use of this function

> thought this might cause some issues

It surely might cause issues but in that case you will get a specific error about tensor not being contagious or having the wrong shape

File: src/models/losses.py (hunk 2 above)
Comment (rohit901) [reply src/models/losses.py:160]: this is a self supervised contrastive objective where the only true matching pair is considered for `Di`, `Qi` where Di and Qi are image and corresponding caption of ith element in batch, all other batch elements are treated as non matching [reference: attnGAN paper]

File: src/models/losses.py [hunk 26]
```diff
@@ -4,6 +4,327 @@
...
+        match_scores_t, match_labels
+    )  # This corresponds to L2_w from attnGAN.
+
+    global_incept_feat = global_incept_feat.unsqueeze(0)  # shape: (1, batch, D)
+    sent_emb = sent_emb.unsqueeze(0)  # shape: (1, batch, D)
+
+    incept_feat_norm = torch.norm(
+        global_incept_feat, 2, dim=2, keepdim=True
+    )  # shape: (1, batch, 1)
+    sent_emb_norm = torch.norm(sent_emb, 2, dim=2, keepdim=True)  # shape: (1, batch, 1)
```
[the remark this answers is not in the document]
Comment (Dmmc123) [reply src/models/losses.py:210]: You attached formula for cosine similarity between two Q and D, it's not the formula for norming at least because norming is for one tensor
image

File: src/models/losses.py (hunk 1 above)
Comment (Dmmc123) [reply src/models/losses.py:147]: > would prefer to do it later

It will save time if next time you do things that are efficient as early as possible. Because you may end up leaving an inefficient code because _"it works right?"_ , then you settle, and then you never fix it. 


For now resolved

File: src/models/losses.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (Dmmc123) [reply src/models/losses.py:108]: then why `prepare_labels` always returns an static arange? maybe if its just an arange it makes sense to eliminate it from `prepare_labels` and always create it on demand? reasons for doing that:
1. you will have one less return value for a function `prepare_labels` 
2. you will have -1 argument in two losses

File: src/models/losses.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:45]: umm but why should the losses part handle the label creation logic? even real labels is just a label of ones and fake labels is label of 0's, but we do pass it in downstream functions... I mean you are right that function notation will have less argument and it can be created easily, but i'm confused like if you tomorrow want to change the logic of match labels you have to change it everywhere where it is used, currently there is just a single source of truth for labels. let me know your thoughts.

File: src/models/losses.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:56]: i see, the tensor is (batch, batch) and we use the operation twice once for masking local damsm features other for global.
for now i have removed device from the loss, so i guess you can verify and resolve.

File: src/models/losses.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:31]: right.. that makes sense. So now we have a trade-off here. Like if we pass directly inception features we lose out on abstraction as the logic would then be in training loop, thus making the training loop bit dirty and unclean.

but if we pass directly the network we will be increasing coupling...
so im not sure how should we optimize here, coupling vs abstraction. let me know what do you think and why it would be better and will try to make the changes accordingly.

[the remark this answers is not in the document]
Comment (Dmmc123) [reply src/models/losses.py:31]: > making the training loop bit dirty and unclean

I would argue that it won't be dirty and unclean. It's just a forward pass through the encoder, which is a natural thing to do within the training loop

File: src/models/losses.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Dmmc123) [reply src/models/losses.py:45]: > even real labels is just a label of ones and fake labels is label of 0's, but we do pass it in downstream functions

that makes sense. resolved.

File: src/models/losses.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:31]: got it, i have made the changes. you can review and resolve once pushed

File: src/models/losses.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:52]: i see, makes sense. i have fixed this too.

File: src/models/losses.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:23]: since i've removed device you can resolve this.

File: src/models/losses.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:13]: you can resolve this as i won't be passing the network to loss now.

File: src/models/losses.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:21]: you can resolve this as vgg won't be passed

File: src/models/losses.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:111]: you can resolve this as i'm not passing device.

File: src/models/losses.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:135]: you can resolve this as device has been removed.

File: src/models/losses.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py:144]: you're right, changed this.

File: src/models/losses.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (rohit901) [reply src/models/losses.py]: we are specifying dimension = 2 argument along which to apply norm so its equivalent to norm for tensor and not matrix from the docs.. i think the formula follows from the paper, if i misunderstood, could you rephrase your question?

[the remark this answers is not in the document]
Comment (Dmmc123) [reply src/models/losses.py]: I completely missed the idea that you implemented cosine similarity from scratch... My new questions is why? PyTorch already has it, you don't have to calculate norms and apply the dot product, just pass it here in a function

source: [pytorch docs](https://pytorch.org/docs/stable/generated/torch.nn.functional.cosine_similarity.html)

File: src/models/losses.py [hunk 27]
```diff
@@ -141,7 +140,7 @@ def damsm_loss(
         mask[i] = 0
         masks.append(mask.reshape(1, -1))  # shape: (1, batch)
 
-        numb_words = cap_len_list[i]
+        numb_words = int(cap_lens[i])
```
Comment (Dmmc123) [inline src/models/losses.py:143]: Do you even need to:

1. Cast it?
2. Make it a separate variable?

To my view, you can just do `query_words = words_emb[i, :, :cap_lens[i]].unsqueeze(0)`

File: src/models/losses.py [hunk 28]
```diff
@@ -187,28 +187,21 @@ def damsm_loss(
...
 
-    global_match_score = global_match_score.squeeze()  # shape: (batch, batch)
+    # mask out the scores for mis-matched samples
     global_match_score.data.masked_fill_(  # type: ignore
         masks, -float("inf")
-    )  # mask out the scores for mis-matched samples
+    )  
 
-    global_match_t = global_match_score.transpose(0, 1)  # shape: (batch, batch)
+    global_match_t = global_match_score.T  # shape: (batch, batch)
```
Comment (Dmmc123) [inline src/models/losses.py:204]: why transpose?

**Review by Dmmc123 — APPROVED** [review]

File: src/models/losses.py (hunk 28 above)
Comment (rohit901) [reply src/models/losses.py:204]: from attnGAN paper, P(D|Q) and P(Q|D) follows the same from word loss calculation.
